### PR TITLE
1409 new price flow split into separate command

### PR DESCRIPTION
**Pull request by jonasdmoeller**
## Description

Introduction of ChargePriceCommand to support new price flow.

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargeCommandReceiptService.cs [hunk 1]
```diff
@@ -43,13 +43,13 @@ public class ChargeCommandReceiptService : IChargeCommandReceiptService
             _acceptedMessageDispatcher = acceptedMessageDispatcher;
         }
 
-        public async Task RejectAsync(ChargeCommand command, ValidationResult validationResult)
+        public async Task RejectAsync(ChargeInformationCommand command, ValidationResult validationResult)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargeCommandReceiptService.cs:46]: Should we rename the class to `ChargeInformationCommandReceiptService` or will it not make sense as long as we still handle prices in the "information"-flow?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandCommandBundleHandler.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+using System;
+using System.Threading.Tasks;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands;
+using GreenEnergyHub.Charges.Domain.Dtos.SharedDtos;
+
+namespace GreenEnergyHub.Charges.Application.Charges.Handlers
+{
+    public class ChargeInformationCommandCommandBundleHandler : IChargeInformationCommandBundleHandler
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandCommandBundleHandler.cs:23]: `CommandCommand`? :-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeInformationCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+// limitations under the License.
+
+using System.Threading.Tasks;
+using System.Transactions;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
+
+namespace GreenEnergyHub.Charges.Application.Charges.Handlers
+{
+    /// <summary>
+    /// Contract for handling a change of charges message.
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeInformationCommandHandler.cs:22]:
```suggestion
    /// Contract for handling a charge information message.
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandBundleHandler.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+// limitations under the License.
+
+using System.Threading.Tasks;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands;
+
+namespace GreenEnergyHub.Charges.Application.Charges.Handlers
+{
+    /// <summary>
+    /// Contract for handling a charge bundle.
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandBundleHandler.cs:22]:
```suggestion
    /// Contract for handling a charge price bundle.
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandBundleHandler.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    /// </summary>
+    public interface IChargePriceCommandBundleHandler
+    {
+        /// <summary>
+        /// Synchronously handle the bundle.
+        /// </summary>
+        /// <param name="informationBundle">Charges bundle</param>
+        Task HandleAsync(ChargeCommandPriceBundle informationBundle);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandBundleHandler.cs:30]:
```suggestion
        /// <param name="priceBundle">Charges bundle</param>
        Task HandleAsync(ChargeCommandPriceBundle priceBundle);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandHandler.cs [hunk 6]
```diff
@@ -14,20 +14,20 @@
 
 using System.Threading.Tasks;
 using System.Transactions;
-using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands;
 
 namespace GreenEnergyHub.Charges.Application.Charges.Handlers
 {
     /// <summary>
     /// Contract for handling a change of charges message.
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandHandler.cs:22]:
```suggestion
    /// Contract for handling a charge price message.
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandHandler.cs [hunk 7]
```diff
@@ -14,20 +14,20 @@
...
-    public interface IChargeCommandHandler
+    public interface IChargePriceCommandHandler
     {
         /// <summary>
         /// Synchronously handle the message.
         /// Supports <see cref="TransactionScope"/>.
         /// </summary>
         /// <param name="command">ChargeCommand</param>
-        Task HandleAsync(ChargeCommand command);
+        Task HandleAsync(ChargePriceCommand command);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargePriceCommandHandler.cs:31]:
```suggestion
        Task HandleAsync(ChargePriceCommand priceCommand);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandHandler.cs [hunk 8]
```diff
@@ -15,20 +15,19 @@
...
 using NodaTime;
 
 namespace GreenEnergyHub.Charges.Application.Charges.Handlers
 {
-    public class ChargeCommandHandler : IChargeCommandHandler
+    public class ChargeInformationCommandHandler : IChargeInformationCommandHandler
     {
         private readonly IClock _clock;
         private readonly IMessageDispatcher<ChargeCommandReceivedEvent> _chargeMessageDispatcher;
         private readonly IMessageDispatcher<ChargePriceCommandReceivedEvent> _chargePriceMessageDispatcher;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandHandler.cs:28]: Let's remove this unused dependency:
```suggestion
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandHandler.cs [hunk 9]
```diff
@@ -15,20 +15,19 @@
...
 
-        public ChargeCommandHandler(
+        public ChargeInformationCommandHandler(
             IClock clock,
             IMessageDispatcher<ChargeCommandReceivedEvent> chargeMessageDispatcher,
             IMessageDispatcher<ChargePriceCommandReceivedEvent> chargePriceMessageDispatcher)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandHandler.cs:33]:
```suggestion
            IMessageDispatcher<ChargeCommandReceivedEvent> chargeMessageDispatcher)
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandHandler.cs [hunk 10]
```diff
@@ -38,18 +37,10 @@ public class ChargeCommandHandler : IChargeCommandHandler
             _chargePriceMessageDispatcher = chargePriceMessageDispatcher;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandHandler.cs:37]:
```suggestion
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs [hunk 11]
```diff
@@ -31,18 +36,24 @@ public class ChargeIngestion
         /// The name of the function.
         /// Function name affects the URL and thus possibly dependent infrastructure.
         /// </summary>
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs:38]: The summary in #35-38 seems to be misplaced, or even irrellevant?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs [hunk 12]
```diff
@@ -51,35 +62,50 @@ public class ChargeIngestion
...
+                        await _chargePriceCommandBundleHandler.HandleAsync(commandBundle).ConfigureAwait(false);
+                        break;
+                }
 
-            return _httpResponseBuilder.CreateAcceptedResponse(req);
+                return _httpResponseBuilder.CreateAcceptedResponse(request);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(exception, "Unable to deserialize request");
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs:100]: Could we add `CorrelationId` to ease troubleshooting?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargePriceCommandReceiverConfiguration.cs [hunk 13]
```diff
@@ -23,7 +23,7 @@ internal static class ChargePriceCommandReceiverConfiguration
     {
         internal static void ConfigureServices(IServiceCollection serviceCollection)
         {
-            serviceCollection.AddScoped<IChargePriceEventHandler, ChargePriceEventHandler>();
+            serviceCollection.AddScoped<IChargePriceEventHandler, ChargePriceEventHandlerDeprecated>();
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargePriceCommandReceiverConfiguration.cs:26]: Not sure, but I think this change is incorrect(?)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandBundleConverter.cs [hunk 14]
```diff
@@ -29,32 +31,137 @@
...
+                case BusinessReasonCode.UpdateChargeInformation:
+                    var chargeOperationsAsync = await ParseChargeOperationsAsync(reader).ConfigureAwait(false);
+                    var chargeCommands = chargeOperationsAsync
+                        .GroupBy(x => new { x.ChargeId, x.ChargeOwner, x.Type })
+                        .Select(chargeOperationDtoGroup =>
+                            new ChargeInformationCommand(
+                                document,
+                                chargeOperationDtoGroup.AsEnumerable().Select(dto => dto).ToList()))
+                        .ToList();
+                    return new ChargeCommandInformationBundle(document, chargeCommands);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandBundleConverter.cs:62]: I suggest we put this in a `CreateChargeCommandInformationBundle()` method, and likewise for the price branch, to enhance readability of the `switch` statement.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeData/AvailableChargeDataFactory.cs [hunk 15]
```diff
@@ -20,7 +20,8 @@
 using GreenEnergyHub.Charges.Core.DateTime;
 using GreenEnergyHub.Charges.Domain.Charges;
 using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommandAcceptedEvents;
-using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
+
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeData/AvailableChargeDataFactory.cs:24]:
```suggestion
using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeReceiptData/AvailableChargeReceiptDataFactory.cs [hunk 16]
```diff
@@ -17,7 +17,8 @@
 using System.Threading.Tasks;
 using GreenEnergyHub.Charges.Application.Messaging;
 using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommandAcceptedEvents;
-using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
+
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeReceiptData/AvailableChargeReceiptDataFactory.cs:21]:
```suggestion
using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeReceiptData/AvailableChargeRejectionDataFactory.cs [hunk 17]
```diff
@@ -18,7 +18,8 @@
 using System.Threading.Tasks;
 using GreenEnergyHub.Charges.Application.Messaging;
 using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommandRejectedEvents;
-using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
+
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeReceiptData/AvailableChargeRejectionDataFactory.cs:22]:
```suggestion
using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Application/ChargesCommandBundleBuilder.cs [hunk 18]
```diff
@@ -13,23 +13,31 @@
...
 
 using System.Collections.Generic;
-using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
+using GreenEnergyHub.Charges.Domain.Dtos.SharedDtos;
 
 namespace GreenEnergyHub.Charges.Tests.Builders.Application
 {
-    public class ChargeCommandBundleBuilder
+    public class ChargeBundleDtoBuilder
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Application/ChargesCommandBundleBuilder.cs:21]: Let's rename the file, too.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeCommandBuilder.cs [hunk 19]
```diff
@@ -48,9 +48,9 @@ public ChargeCommandBuilder WithChargeOperations(List<ChargeOperationDto> charge
             return this;
         }
 
-        public ChargeCommand Build()
+        public ChargeInformationCommand Build()
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeCommandBuilder.cs:51]: Let's rename the file, too.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure.Core/MessagingExtensions/Serialization/JsonMessageDeserializerTests.cs [hunk 20]
```diff
@@ -47,6 +49,11 @@ public class JsonMessageDeserializerTests
         public async Task FromBytesAsync_CreatesMessage(IInboundMessage expected)
         {
             // Arrange
+            if (expected is ChargeCommandInformationBundle)
+            {
+                Debugger.Break();
+            }
+
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure.Core/MessagingExtensions/Serialization/JsonMessageDeserializerTests.cs:56]: Remove before merge?

**Review by HenrikSommer — COMMENTED** [review]
I've commented here and there, but overall it's a great job!

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargeCommandReceiptService.cs (hunk 1 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargeCommandReceiptService.cs:46]: I think we should consider reusing the reciept service as a generic solution for both charge information and price flows - both "old" and "new".

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandCommandBundleHandler.cs (hunk 2 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationCommandCommandBundleHandler.cs:23]: Fixed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargePriceCommandReceiverConfiguration.cs (hunk 13 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargePriceCommandReceiverConfiguration.cs:26]: This is to support a new ChargePriceEventHandler to be injected into the new ChargePriceCommandReceiverEndpoint.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandBundleConverter.cs (hunk 14 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandBundleConverter.cs:62]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandBundleConverter.cs [hunk 21]
```diff
@@ -29,32 +31,137 @@
...
+                if (reader.Is(CimChargeCommandConstants.Id))
+                {
+                    var content = await reader.ReadValueAsStringAsync().ConfigureAwait(false);
+                    operationId = content;
+                }
+                else if (reader.Is(CimChargeCommandConstants.ChargeGroup))
+                {
+                    while (await reader.AdvanceAsync().ConfigureAwait(false))
+                    {
+                        if (reader.Is(CimChargeCommandConstants.ChargeTypeElement))
```
[the remark this answers is not in the document]
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandBundleConverter.cs:96]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs (hunk 12 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs:100]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeCommandBuilder.cs (hunk 19 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeCommandBuilder.cs:51]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure.Core/MessagingExtensions/Serialization/JsonMessageDeserializerTests.cs (hunk 20 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure.Core/MessagingExtensions/Serialization/JsonMessageDeserializerTests.cs:56]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargePriceCommandReceiverConfiguration.cs (hunk 13 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargePriceCommandReceiverConfiguration.cs:26]: Yes, but in my mind, the `ChargePriceEventHandler` should be part of this PR, but that is not important.

**Review by HenrikSommer — APPROVED** [review]

**Pull request merged**